File: Algorithm/107. Binary Tree Level Order Traversal II.py
# -*- coding: utf-8 -*-
# @Author: WuLC
# @Date:   2016-06-14 14:42:14
# @Last modified by:   WuLC
# @Last Modified time: 2016-06-14 14:42:22
# @Email: liangchaowu5@gmail.com

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def levelOrderBottom(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        if root == None: return []
        nodes, result= [root], []
        while len(nodes) > 0:
            next_level, tmp = [], []
            for i in xrange(len(nodes)):
                tmp.append(nodes[i].val)
                if nodes[i].left:
                    next_level.append(nodes[i].left)
                if nodes[i].right:
                    next_level.append(nodes[i].right)
            result.append(tmp)
            nodes = next_level
        r = []
        while len(result)>0:
            r.append(result.pop())
        return r

File: Algorithm/Java/107. Binary Tree Level Order Traversal II.java
/*
* @Author: WuLC
* @Date:   2017-07-31 16:00:19
* @Last Modified by:   WuLC
* @Last Modified time: 2017-07-31 16:01:01
* @Email: liangchaowu5@gmail.com
*/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

// traverse level by level
public class Solution 
{
    public List<List<Integer>> levelOrderBottom(TreeNode root) 
    {
        List<List<Integer>> result = new ArrayList<List<Integer>>();
        if(root == null) return result;
        List<Integer> tmp;
        List<TreeNode> curr, next;
        next = new ArrayList<TreeNode>();
        next.add(root);
        while(next.size() > 0)
        {
            curr = next;
            next = new ArrayList<TreeNode>();
            tmp =  new ArrayList<Integer>();
            for(int i = 0; i < curr.size(); i++)
            {
                tmp.add(curr.get(i).val);
                if(curr.get(i).left != null)  next.add(curr.get(i).left);
                if(curr.get(i).right != null) next.add(curr.get(i).right);
            }  
            result.add(tmp);
        }
        Collections.reverse(result);
        return result;
    }
}